kubeca/certinit/certreq.go

```diff
@@ -0,0 +1,166 @@
+package certinit
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/ekspand/trusty/pkg/csr"
+	"github.com/ekspand/trusty/pkg/inmemcrypto"
+	"github.com/juju/errors"
+	capi "k8s.io/api/certificates/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	randAlphabet = "0123456789abcdefghijklmnopqrstuvwxyz"
+)
+
+var (
+	randAlphabetLength = big.NewInt(int64(len(randAlphabet)))
+)
+
+var usages = map[string][]capi.KeyUsage{
+	"peer": {
+		capi.UsageDigitalSignature,
+		capi.UsageKeyEncipherment,
+		capi.UsageServerAuth,
+		capi.UsageClientAuth,
+	},
+	"client": {
+		capi.UsageDigitalSignature,
+		capi.UsageKeyEncipherment,
+		capi.UsageClientAuth,
+	},
+	"server": {
+		capi.UsageDigitalSignature,
+		capi.UsageKeyEncipherment,
+		capi.UsageServerAuth,
+	},
+}
+
+func (r *Request) requestCertificate(ctx context.Context, client MinCertificates) error {
+	issuerAndProfile := strings.Split(r.SignerName, "/")
+	if len(issuerAndProfile) != 2 {
+		return errors.Errorf("unsupported signer: " + r.SignerName)
+	}
+
+	profileUsages := usages[issuerAndProfile[1]]
+	if profileUsages == nil {
+		return errors.Errorf("unsupported profile: " + r.SignerName)
+	}
+
+	prov := csr.NewProvider(inmemcrypto.NewProvider())
+	req := csr.CertificateRequest{
+		KeyRequest: prov.NewKeyRequest("", "ECDSA", 256, csr.SigningKey),
+		SAN:        r.san,
+	}
+
+	pemCsr, pemKeyBytes, _, _, err := prov.CreateRequestAndExportKey(&req)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	keyFile := path.Join(r.CertDir, "tls.key")
+	if err := ioutil.WriteFile(keyFile, pemKeyBytes, 0600); err != nil {
+		return errors.Annotate(err, "unable to safe key")
```

Comment:
saVe typo... also you might want to add keyFile to the error message for the context.